src/ConnectedMode/QualityProfiles/OutOfDateQPFinder.cs

```diff
@@ -0,0 +1,115 @@
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarLint.VisualStudio.Integration;
+using SonarQube.Client;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.QualityProfiles
+{
+    internal interface IOutOfDateQPFinder
+    {
+        /// <summary>
+        /// When in Connected Mode, returns the languages for which the Quality Profile
+        /// needs to be updated
+        /// </summary>
+        Task<IEnumerable<Language>> GetAsync(CancellationToken cancellationToken);
+    }
+
+    [Export(typeof(IOutOfDateQPFinder))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class OutOfDateQPFinder : IOutOfDateQPFinder
+    {
+        private readonly IConfigurationProvider configProvider;
+        private readonly ISonarQubeService sonarQubeService;
+        private readonly ILogger logger;
+
+        [ImportingConstructor]
+        public OutOfDateQPFinder(IConfigurationProvider configProvider,
+            ISonarQubeService sonarQubeService,
+            ILogger logger)
+        {
+            this.configProvider = configProvider;
+            this.sonarQubeService = sonarQubeService;
+            this.logger = logger;
+        }
+
+        public async Task<IEnumerable<Language>> GetAsync(CancellationToken cancellationToken)
+        {
+            var languagesToUpdate = new List<Language>();
+
+            var config = configProvider.GetConfiguration();
+            foreach (var profile in config.Project.Profiles)
+            {
+                LogQPVerbose(profile.Key, "Checking if Quality Profile has changed");
+                if (await IsUpdateRequiredAsync(config, profile, cancellationToken))
+                {
+                    languagesToUpdate.Add(profile.Key);
+                    LogQPVerbose(profile.Key, "Quality profile update is required");
+                }
+                else
+                {
+                    LogQPVerbose(profile.Key, "Quality profile is up-to-date");
+                }
+            }
+
+            return languagesToUpdate;
+        }
+
+        private async Task<bool> IsUpdateRequiredAsync(BindingConfiguration config, KeyValuePair<Language, ApplicableQualityProfile> profile, CancellationToken cancellationToken)
+        {
+            Debug.Assert(profile.Value != null, "Not expecting profile to be null. Language: " + profile.Key);
+
+            var serverProfileInfo = await sonarQubeService.GetQualityProfileAsync(config.Project.ProjectKey,
+                config.Project.Organization?.Key,
+                profile.Key.ServerLanguage, CancellationToken.None);
+
+            return HasProfileChanged(serverProfileInfo, profile.Value);
+        }
+
+        private bool HasProfileChanged(SonarQubeQualityProfile newProfileInfo, ApplicableQualityProfile oldProfileInfo)
```

Comment:
There used to be a component that did a similar thing that was part of the now-deleted `ErrorListInfoBarController`. This method uses the same logic as [the old code](https://github.com/SonarSource/sonarlint-visualstudio/blob/75f3826e72a584818070801a1d5c7808cb4a8aa0/src/Integration/LocalServices/ErrorListInfoBarController.cs#L741).